text/0000-proc-macros.md

```diff
@@ -0,0 +1,417 @@
+compiler to be practical).
+
+We could have a dedicated syntax for procedural macros, similar to the
+`macro_rules` syntax for macros by example. Since a procedural macro is really
+just a Rust function, I believe using a function is better. I have also not been
+able to come up with (or seen suggestions for) a good alternative syntax. It
+seems reasonable to expect to write Rust macros in Rust (although there is
+nothing stopping a macro author from using FFI and some other language to write
+part or all of a macro).
+
+For attribute-like macros on items, it would be nice if we could skip parsing
+the annotated item until after macro expansion. That would allow for more
+flexible macros, since the input would not be constrained to Rust syntax. However,
+this would require identifying items from tokens, rather than from the AST, which
+would require additional rules on token trees and may not be possible.
+
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+### macros with an extra identifier
+
+We currently allow procedural macros to take an extra ident after the macro name
+and before the arguments, e.g., `foo! bar(...)` where `foo` is the macro name
+and `bar` is the extra identifier. This is used for `macro_rules` and is useful
+for macros which define classes of items, rather than instances of items. E.g.,
+a `struct!` macro might be used similarly to the `struct` keyword.
+
+My feeling is that this macro form is not used enough to justify its existence.
+From a design perspective, it encourages uses of macros for language extension,
+rather than syntactic abstraction. I feel that such macros are at higher risk of
+making programs incomprehensible and of fragmenting the ecosystem).
+
+Therefore, I would like to remove them from the language. Alternatively, they
+could be incorporated into the new design by having another kind of macro
+function:
+
+```
+#[macro_with_ident]
+pub fn foo(&Token, TokenStream, &mut MacroContext) -> TokenStream;
+```
+
+where the first argument is the extra identifier.
+
+
+### Linking model
+
+Currently, procedural macros are dynamically linked with the compiler. This
+prevents the compiler being statically linked, which is sometimes desirable. An
+alternative architecture would have procedural macros compiled as independent
+programs and have them communicate with the compiler via IPC.
+
+This would have the advantage of allowing static linking for the compiler and
+would prevent procedural macros from crashing the main compiler process.
+However, designing a good IPC interface is complicated because there is a lot of
+data that might be exchanged between the compiler and the macro.
+
+I think we could first design the syntax, interfaces, etc. and later evolve into
+a process-separated model (if desired). However, if this is considered an
+essential feature of macro reform, then we might want to consider the interfaces
+more thoroughly with this in mind.
```

Comment:
We might have to do clever things like use shared memory alongside message-based IPC to get the most out of such a model, but I believe it would be worth it.

Dynamic linking for plugin systems works in practice but just making it safe to unload said plugins is an entire area of design Rust hasn't even touched yet (although lifetimes would play a very important role).

I would rather have an IPC solution that lets me:
* download a *single* binary and have a working cross-compiler without even system dependencies (static linking against musl on linux)
* safely unload plugins after expansion without introducing complexity into the language
* use a forking zygote to share most of the plugins across compilations (via cargo or RLS)
* implement my own libmacro in an alternative Rust compiler
* more exciting: implement my own libmacro for a Rust tool which is not a full compiler